src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,548 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! [KvBackend] implementation based on [raft_engine::Engine].
+
+use std::any::Any;
+use std::sync::RwLock;
+
+use common_meta::kv_backend::txn::{Txn, TxnOp, TxnOpResponse, TxnRequest, TxnResponse};
+use common_meta::kv_backend::{KvBackend, TxnService};
+use common_meta::rpc::store::{
+    BatchDeleteRequest, BatchDeleteResponse, BatchGetRequest, BatchGetResponse, BatchPutRequest,
+    BatchPutResponse, CompareAndPutRequest, CompareAndPutResponse, DeleteRangeRequest,
+    DeleteRangeResponse, MoveValueRequest, MoveValueResponse, PutRequest, PutResponse,
+    RangeRequest, RangeResponse,
+};
+use common_meta::rpc::KeyValue;
+use raft_engine::{Engine, LogBatch};
+use snafu::ResultExt;
+
+use crate::error;
+use crate::error::RaftEngineSnafu;
+
+pub(crate) const SYSTEM_NAMESPACE: u64 = 0;
+
+/// RaftEngine based [KvBackend] implementation.
+pub struct RaftEngineBackend {
+    engine: RwLock<Engine>,
+}
+
+#[async_trait::async_trait]
+impl TxnService for RaftEngineBackend {
+    type Error = error::Error;
+
+    async fn txn(&self, txn: Txn) -> Result<TxnResponse, Self::Error> {
+        let TxnRequest {
+            compare,
+            success,
+            failure,
+        } = txn.into();
+
+        let mut succeeded = true;
+        let engine = self.engine.write().unwrap();
+        for cmp in compare {
+            let existing_value = engine_get(&engine, &cmp.key)?.map(|kv| kv.value);
+            if !cmp.compare_with_value(existing_value.as_ref()) {
+                succeeded = false;
+                break;
+            }
+        }
+
+        let mut batch = LogBatch::default();
+        let do_txn = |txn_op| match txn_op {
+            TxnOp::Put(key, value) => {
+                batch
+                    .put(SYSTEM_NAMESPACE, key.clone(), value)
+                    .context(RaftEngineSnafu)?;
+                Ok(TxnOpResponse::ResponsePut(PutResponse { prev_kv: None }))
+            }
+
+            TxnOp::Get(key) => {
+                let value = engine_get(&engine, &key)?.map(|kv| kv.value);
+                let kvs = value
+                    .into_iter()
+                    .map(|value| KeyValue {
+                        key: key.clone(),
+                        value,
+                    })
+                    .collect();
+                Ok(TxnOpResponse::ResponseGet(RangeResponse {
+                    kvs,
+                    more: false,
+                }))
+            }
+
+            TxnOp::Delete(key) => {
+                let prev_value = engine_get(&engine, &key)?;
+                batch.delete(SYSTEM_NAMESPACE, key);
+
+                let prev_kvs: Vec<_> = prev_value.into_iter().collect();
```

Comment:
Let's maintain consistent behavior and keep prev_kvs empty. Similarly, please help fix [memoryKv](https://github.com/GreptimeTeam/greptimedb/blob/46fa3eb6293931700d5ec1feeb65a2c962d89653/src/common/meta/src/kv_backend/memory.rs#L342). ^_^

Reply:
fixed in [257c7a6](https://github.com/GreptimeTeam/greptimedb/pull/2280/commits/257c7a6d3cd42ac501dfecfa3533502c5011f0e9#diff-7a0072870cc873b1a630e534dde51ccc8032ded0ad3d8581a17b38e62a9b8b8cR340-R343)